dev-tools/omdb/src/bin/omdb/mgs.rs

```diff
@@ -0,0 +1,499 @@
+                let addr = addrs.into_iter().next().expect(
+                    "expected at least one MGS address from \
+                    successful DNS lookup",
+                );
+                format!("http://{}", addr)
+            }
+        };
+        eprintln!("note: using MGS URL {}", &mgs_url);
+        let mgs_client = gateway_client::Client::new(&mgs_url, log.clone());
+
+        match &self.command {
+            MgsCommands::Inventory(inventory_args) => {
+                cmd_mgs_inventory(&mgs_client, inventory_args).await
+            }
+        }
+    }
+}
+
+/// Runs `omdb mgs inventory`
+///
+/// Shows devices and components that are visible to an MGS instance.
+async fn cmd_mgs_inventory(
+    mgs_client: &gateway_client::Client,
+    _args: &InventoryArgs,
+) -> Result<(), anyhow::Error> {
+    // Report all the SP identifiers that MGS is configured to talk to.
+    println!("ALL CONFIGURED SPs\n");
+    let mut sp_ids = mgs_client
+        .sp_all_ids()
+        .await
+        .context("listing SP identifiers")?
+        .into_inner();
+    sp_ids.sort_by(sp_id_cmp);
+    show_sp_ids(&sp_ids)?;
+    println!("");
+
+    // Report which SPs are visible via Ignition.
+    println!("SPs FOUND THROUGH IGNITION\n");
+    let mut sp_list_ignition = mgs_client
+        .ignition_list()
+        .await
+        .context("listing ignition")?
+        .into_inner();
+    sp_list_ignition.sort_by(|a, b| sp_id_cmp(&a.id, &b.id));
+    show_sps_from_ignition(&sp_list_ignition)?;
+    println!("");
+
+    // Print basic state about each SP that's visible to ignition.
+    println!("SERVICE PROCESSOR STATES\n");
+    let mgs_client = std::sync::Arc::new(mgs_client);
+    let c = &mgs_client;
+    let mut sp_infos =
+        futures::stream::iter(sp_list_ignition.iter().filter_map(|ignition| {
+            if matches!(ignition.details, SpIgnition::Yes { .. }) {
+                Some(ignition.id)
+            } else {
+                None
+            }
+        }))
+        .then(|sp_id| async move {
+            let msg = format!("fetching info about SP {:?}", sp_id);
+            c.sp_get(sp_id.type_, sp_id.slot)
+                .await
+                .context(msg)
+                .map(|s| (sp_id, s))
+        })
+        .collect::<Vec<Result<_, _>>>()
+        .await
+        .into_iter()
+        .filter_map(|r| match r {
+            Ok((sp_id, v)) => Some((sp_id, v.into_inner())),
+            Err(error) => {
+                eprintln!("error: {:?}", error);
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+    sp_infos.sort_by(|s1, s2| sp_id_cmp(&s1.0, &s2.0));
+    show_sp_states(&sp_infos)?;
+    println!("");
+
+    // Print detailed information about each SP that we've found so far.
+    for (sp_id, sp_state) in &sp_infos {
+        show_sp_details(&mgs_client, sp_id, sp_state).await?;
+    }
+
+    Ok(())
+}
+
+fn sp_type_to_str(s: &SpType) -> &'static str {
```

Comment:
This is fine, but stylistically I probably would've written this as an extension trait to make the call sites nicer:

```rust
trait SpTypeExt {
    fn as_str(&self) -> &'static str;
}

impl SpTypeExt for SpType { .. }
```